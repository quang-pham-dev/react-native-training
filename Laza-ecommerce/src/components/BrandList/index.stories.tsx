import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import React from 'react';
import CenterView from '../../../storybook/stories/CenterView';
import BrandCardList from './index';
import { brands } from '__mocks__/dataMock/brands';

storiesOf('Brand List', module)
  .addDecorator(getStory => <CenterView>{getStory()}</CenterView>)
  .add('Default', () => (
    <BrandCardList
      brands={brands}
      onPressBrandCard={action('clicked')}
      onLoadMoreBrands={action('clicked')}
    />
  ));
