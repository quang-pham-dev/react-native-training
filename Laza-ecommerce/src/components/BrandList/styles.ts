import { StyleSheet } from 'react-native';
import { Metrics } from 'styles/themes';
const styles = StyleSheet.create({
  container: {},
  brandTitle: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingBottom: Metrics.padding.xm,
  },
});

export default styles;
