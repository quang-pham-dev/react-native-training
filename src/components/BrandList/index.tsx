import React, {memo, useCallback} from 'react'
import {FlatList} from 'react-native'

// LIBS
import isEqual from 'react-fast-compare'

// Components
import BrandCard from '@components/BrandCard'
import LoadingIndicator from '@components/LoadingIndicator'

// Constants
import {BRANDS_EMPTY_RESULT} from '@constants'

// Types
import {BrandCardListProps, BrandsListProps} from '@model-types'

// Styles
import PStyled from '@components/Paragraph/P.styles'
import ViewStyled from '@components/View/View.styles'

const BrandList = ({onPressBrandCard, brands}: BrandsListProps) => {
  // handle action when press card brand with id
  const handlePressBrandCard = useCallback(
    (id: string) => {
      onPressBrandCard(id)
    },
    [onPressBrandCard],
  )

  // handle action load more brands
  const handleLoadMoreBrands = () => {}

  // handle render Card component
  const renderBrandCard = ({item}: {item: BrandCardListProps}) => (
    <BrandCard brand={item} onPressBrandCard={handlePressBrandCard} />
  )

  // render empty list
  const renderEmptyList = () => (
    <PStyled.Base>{BRANDS_EMPTY_RESULT}</PStyled.Base>
  )

  // handle render Footer component
  const renderFooterComponent = () => <LoadingIndicator />

  return (
    <ViewStyled.Default>
      <FlatList
        horizontal
        data={brands}
        renderItem={renderBrandCard}
        keyExtractor={brand => brand.id}
        showsHorizontalScrollIndicator={false}
        ListEmptyComponent={renderEmptyList}
        ListFooterComponent={renderFooterComponent}
        onEndReached={handleLoadMoreBrands}
        onEndReachedThreshold={0.5}
      />
    </ViewStyled.Default>
  )
}

export default memo(BrandList, isEqual)
