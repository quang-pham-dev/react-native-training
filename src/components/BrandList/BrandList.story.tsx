import React from 'react'
import {action} from '@storybook/addon-actions'
import {storiesOf} from '@storybook/react-native'
import {StoryScreen, Story, UseCase} from '../../../storybook/views'

import BrandList from './index'

import {brands} from '@constants/dataMock'

declare let module: NodeModule

storiesOf('Brand List', module)
  .addDecorator(fn => <StoryScreen>{fn()}</StoryScreen>)
  .add('Default', () => (
    <Story>
      <UseCase text="default">
        <BrandList
          brands={brands}
          onPressBrandCard={action('clicked')}
          onLoadMoreBrand={action('clicked')}
        />
      </UseCase>
    </Story>
  ))
