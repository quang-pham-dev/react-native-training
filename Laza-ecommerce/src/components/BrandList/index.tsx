import React, { memo, useCallback, useContext } from 'react';
import { FlatList, View } from 'react-native';

// LIBS
import isEqual from 'react-fast-compare';

// Context
import { BrandsContext } from 'context/BrandsContext';

// Components
import BrandCard from 'components/BrandCard';
import LoadingIndicator from 'components/LoadingIndicator';
import Title from 'components/Title';

// Constants
import { BRANDS_EMPTY_RESULT } from 'constants/Brands';

// Types
import { IBrandCardListProps, IBrandsListProps } from 'types/models/Brands';

// Styles
import styles from './styles';

const BrandsCardList = ({ onPressBrandCard, brands, onLoadMoreBrands }: IBrandsListProps) => {
  const { brandState } = useContext(BrandsContext);

  const { totalRowsOfBrands, isLoading } = brandState || {};

  // handle action when press card brand with id
  const handlePressBrandCard = useCallback(
    (id: string) => {
      onPressBrandCard(id);
    },
    [onPressBrandCard],
  );

  // handle action load more brands
  const handleLoadMoreBrands = () => {
    let cacheEndReached = null;
    brands?.length < totalRowsOfBrands && (cacheEndReached = onLoadMoreBrands);

    return cacheEndReached;
  };

  // handle render Card component
  const renderBrandCard = ({ item }: { item: IBrandCardListProps }) => (
    <BrandCard brand={item} onPressBrandCard={handlePressBrandCard} />
  );

  // render empty list
  const renderEmptyList = () => <Title titleName={BRANDS_EMPTY_RESULT} />;

  // handle render Footer component
  const renderFooterComponent = () => {
    {
      isLoading && <LoadingIndicator />;
    }
    return null;
  };

  return (
    <View style={styles.container}>
      <FlatList
        data={brands}
        renderItem={renderBrandCard}
        keyExtractor={brand => brand.id}
        horizontal={true}
        showsHorizontalScrollIndicator={false}
        ListEmptyComponent={renderEmptyList}
        ListFooterComponent={renderFooterComponent}
        onEndReached={handleLoadMoreBrands()}
        onEndReachedThreshold={0.5}
      />
    </View>
  );
};

export default memo(BrandsCardList, isEqual);
