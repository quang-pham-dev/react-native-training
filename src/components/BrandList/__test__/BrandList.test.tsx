import React from 'react'

// LIBS
import {render} from '@testing-library/react-native'

// Components
import BrandList from '@components/BrandList'

// Mocks
import {brands} from '@constants'

// Constants
import {BRANDS_EMPTY_RESULT} from '@constants'

describe('Brand List Component', () => {
  const props = {
    onPressBrandCard: jest.fn(),
    brands: brands,
  }

  const tree = render(<BrandList {...props} />)
  test('Should render correctly', () => {
    const component = tree.toJSON()
    expect(component).toMatchSnapshot()
  })

  test('Should render No Brand found text when brand list is empty', () => {
    const newProps = {
      ...props,
      brands: [],
    }
    const component = render(<BrandList {...newProps} />)
    const {getByText} = component
    const textContent = getByText(BRANDS_EMPTY_RESULT)
    expect(textContent).toBeTruthy()
  })
})
