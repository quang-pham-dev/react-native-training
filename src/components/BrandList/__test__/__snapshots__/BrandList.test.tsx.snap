// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Brand List Component Should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "width": "100%",
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "id": "0",
          "name": "Nike",
          "url": "https://1000logos.net/wp-content/uploads/2021/11/Nike-Logo.png",
        },
        Object {
          "id": "1",
          "name": "Adidas",
          "url": "https://1000logos.net/wp-content/uploads/2016/10/Adidas-logo.png",
        },
        Object {
          "id": "2",
          "name": "Fila",
          "url": "https://1000logos.net/wp-content/uploads/2021/05/Fila-logo.png",
        },
        Object {
          "id": "3",
          "name": "Puma",
          "url": "https://1000logos.net/wp-content/uploads/2021/04/Puma-logo.png",
        },
        Object {
          "id": "4",
          "name": "Gucci",
          "url": "https://1000logos.net/wp-content/uploads/2017/01/Gucci-Logo.png",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#F5F6FA",
              "borderRadius": 10,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
              "opacity": 1,
              "paddingRight": 15,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "marginLeft": 5,
                    "marginRight": 10,
                    "marginVertical": 5,
                    "paddingBottom": 6,
                    "paddingLeft": 6,
                    "paddingRight": 6,
                    "paddingTop": 6,
                  },
                ]
              }
            >
              <Image
                alignSelf="center"
                height={40}
                resizeMode="contain"
                source={
                  Object {
                    "priority": "normal",
                    "uri": "https://1000logos.net/wp-content/uploads/2021/11/Nike-Logo.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "height": 40,
                      "resizeMode": "contain",
                      "width": 40,
                    },
                  ]
                }
                width={40}
              />
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#1D1E20",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 16.5,
                    "textAlign": "left",
                  },
                ]
              }
              type="brand"
            >
              Nike
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#F5F6FA",
              "borderRadius": 10,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
              "opacity": 1,
              "paddingRight": 15,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "marginLeft": 5,
                    "marginRight": 10,
                    "marginVertical": 5,
                    "paddingBottom": 6,
                    "paddingLeft": 6,
                    "paddingRight": 6,
                    "paddingTop": 6,
                  },
                ]
              }
            >
              <Image
                alignSelf="center"
                height={40}
                resizeMode="contain"
                source={
                  Object {
                    "priority": "normal",
                    "uri": "https://1000logos.net/wp-content/uploads/2016/10/Adidas-logo.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "height": 40,
                      "resizeMode": "contain",
                      "width": 40,
                    },
                  ]
                }
                width={40}
              />
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#1D1E20",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 16.5,
                    "textAlign": "left",
                  },
                ]
              }
              type="brand"
            >
              Adidas
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#F5F6FA",
              "borderRadius": 10,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
              "opacity": 1,
              "paddingRight": 15,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "marginLeft": 5,
                    "marginRight": 10,
                    "marginVertical": 5,
                    "paddingBottom": 6,
                    "paddingLeft": 6,
                    "paddingRight": 6,
                    "paddingTop": 6,
                  },
                ]
              }
            >
              <Image
                alignSelf="center"
                height={40}
                resizeMode="contain"
                source={
                  Object {
                    "priority": "normal",
                    "uri": "https://1000logos.net/wp-content/uploads/2021/05/Fila-logo.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "height": 40,
                      "resizeMode": "contain",
                      "width": 40,
                    },
                  ]
                }
                width={40}
              />
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#1D1E20",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 16.5,
                    "textAlign": "left",
                  },
                ]
              }
              type="brand"
            >
              Fila
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#F5F6FA",
              "borderRadius": 10,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
              "opacity": 1,
              "paddingRight": 15,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "marginLeft": 5,
                    "marginRight": 10,
                    "marginVertical": 5,
                    "paddingBottom": 6,
                    "paddingLeft": 6,
                    "paddingRight": 6,
                    "paddingTop": 6,
                  },
                ]
              }
            >
              <Image
                alignSelf="center"
                height={40}
                resizeMode="contain"
                source={
                  Object {
                    "priority": "normal",
                    "uri": "https://1000logos.net/wp-content/uploads/2021/04/Puma-logo.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "height": 40,
                      "resizeMode": "contain",
                      "width": 40,
                    },
                  ]
                }
                width={40}
              />
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#1D1E20",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 16.5,
                    "textAlign": "left",
                  },
                ]
              }
              type="brand"
            >
              Puma
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#F5F6FA",
              "borderRadius": 10,
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
              "opacity": 1,
              "paddingRight": 15,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "marginLeft": 5,
                    "marginRight": 10,
                    "marginVertical": 5,
                    "paddingBottom": 6,
                    "paddingLeft": 6,
                    "paddingRight": 6,
                    "paddingTop": 6,
                  },
                ]
              }
            >
              <Image
                alignSelf="center"
                height={40}
                resizeMode="contain"
                source={
                  Object {
                    "priority": "normal",
                    "uri": "https://1000logos.net/wp-content/uploads/2017/01/Gucci-Logo.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "height": 40,
                      "resizeMode": "contain",
                      "width": 40,
                    },
                  ]
                }
                width={40}
              />
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#1D1E20",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 15,
                    "lineHeight": 16.5,
                    "textAlign": "left",
                  },
                ]
              }
              type="brand"
            >
              Gucci
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Array [
              Object {
                "width": "100%",
              },
            ]
          }
        >
          <ActivityIndicator
            color="#FFFFFF"
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
